sslr-core/src/main/java/com/sonar/sslr/api/typed/ActionParser.java

```diff
@@ -0,0 +1,378 @@
+  }
+
+  public GrammarRuleKey rootRule() {
+    return rootRule;
+  }
+
+  public static class GrammarBuilderInterceptor<N, T> implements MethodInterceptor, GrammarBuilder<N, T>, NonterminalBuilder {
+
+    private final LexerlessGrammarBuilder b;
+    private final BiMap<Method, GrammarRuleKey> mapping = HashBiMap.create();
+    private final BiMap<Method, GrammarRuleKey> actions = HashBiMap.create();
+    private final Set<GrammarRuleKey> optionals = Sets.newHashSet();
+    private final Set<GrammarRuleKey> oneOrMores = Sets.newHashSet();
+    private final Set<GrammarRuleKey> zeroOrMores = Sets.newHashSet();
+
+    private Method buildingMethod = null;
+    private GrammarRuleKey ruleKey = null;
+    private final Deque<ParsingExpression> expressionStack = new ArrayDeque<ParsingExpression>();
+
+    public GrammarBuilderInterceptor(LexerlessGrammarBuilder b) {
+      this.b = b;
+    }
+
+    @Override
+    public Object intercept(Object obj, Method method, Object[] args, MethodProxy proxy) throws Throwable {
+      if (method.getDeclaringClass().equals(Object.class)) {
+        return proxy.invokeSuper(obj, args);
+      }
+
+      if (buildingMethod != null) {
+        push(new DelayedRuleInvocationExpression(b, this, method));
+        return null;
+      }
+
+      buildingMethod = method;
+
+      return proxy.invokeSuper(obj, args);
+    }
+
+    @Override
+    public <U> NonterminalBuilder<U> nonterminal() {
+      return nonterminal(new DummyGrammarRuleKey(this.buildingMethod));
+    }
+
+    @Override
+    public <U> NonterminalBuilder<U> nonterminal(GrammarRuleKey ruleKey) {
+      this.ruleKey = ruleKey;
+      this.mapping.put(this.buildingMethod, this.ruleKey);
+      return this;
+    }
+
+    @Override
+    public Object is(Object method) {
+      Preconditions.checkState(expressionStack.size() == 1, "Unexpected stack size: " + expressionStack.size());
+
+      ParsingExpression expression = pop();
+      b.rule(ruleKey).is(expression);
+
+      this.buildingMethod = null;
+      this.ruleKey = null;
+
+      return null;
+    }
+
+    @Override
+    public <U> U firstOf(U... methods) {
+      ParsingExpression expression = new FirstOfExpression(pop(methods.length));
+      expressionStack.push(expression);
+      return null;
+    }
+
+    @Override
+    public <U> Optional<U> optional(U method) {
+      ParsingExpression expression = pop();
+      GrammarRuleKey grammarRuleKey = new DummyGrammarRuleKey("optional", expression);
+      optionals.add(grammarRuleKey);
+      b.rule(grammarRuleKey).is(b.optional(expression));
+      invokeRule(grammarRuleKey);
+      return null;
+    }
+
+    @Override
+    public <U> List<U> oneOrMore(U method) {
+      ParsingExpression expression = pop();
+      GrammarRuleKey grammarRuleKey = new DummyGrammarRuleKey("oneOrMore", expression);
+      oneOrMores.add(grammarRuleKey);
+      b.rule(grammarRuleKey).is(b.oneOrMore(expression));
+      invokeRule(grammarRuleKey);
+      return null;
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Return an empty collection instead of null. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1168)